sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+    if (psa_key_lifetime_is_external(slot->attr.lifetime)) {
+        status = psa_encode_protected_key_slot(&enc, slot);
+        if (status != PSA_SUCCESS) {
+            return status;
+        }
+        goto done;
+    }
+#endif /* PSA_PROTECTED_KEY_COUNT */
+
+#if PSA_ASYMMETRIC_KEYPAIR_COUNT
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        status = psa_encode_asymmetric_key_pair(&enc, slot);
+        if (status != PSA_SUCCESS) {
+            return status;
+        }
+        goto done;
+    }
+#endif /* PSA_ASYMMETRIC_KEYPAIR_COUNT */
+
+#if PSA_SINGLE_KEY_COUNT
+    status = psa_encode_single_key(&enc, slot);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+    goto done;
+#endif /* PSA_SINGLE_KEY_COUNT */
+
+done:
+    nanocbor_fmt_end_indefinite(&enc);
+    *output_size = nanocbor_encoded_len(&enc);
+
+#if ENABLE_DEBUG
+    printf("CBOR Encoded Length: %d\n", *output_size);
+    puts("CBOR Encoded Key Slot:\n");
+    for (size_t i = 0; i < output_len; i++){
+        printf("%02x ", output[i]);
+    }
+    printf("\n");
+#endif
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while encoding key slot: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+
+/**
+ * @brief   Decode CBOR encoded key data
+ *
+ * @param   key             Nanocbor value object containing CBOR encoding of the key data
+ * @param   key_data        Output buffer to write key value
+ * @param   key_len         Pointer to write actual key size
+ * @return  psa_status_t
+ */
+static psa_status_t psa_decode_key_data(nanocbor_value_t *key, uint8_t *key_data, size_t *key_len)
+{
+    int res;
+
+    res = nanocbor_get_uint32(key, (uint32_t *)key_len);
+    if (res < 0) {
+        goto error;
+    }
+
+    nanocbor_value_t cbor_key_data;
+    res = nanocbor_enter_array(key, &cbor_key_data);
+    if (res < 0) {
+        goto error;
+    }
+
+    size_t pos = 0;
+    while (!nanocbor_at_end(&cbor_key_data)) {
+        res = nanocbor_get_uint8(&cbor_key_data, &key_data[pos]);
+        if (res < 0) {
+            goto error;
+        }
+
+        pos += 1;
+    }
+
+    return PSA_SUCCESS;
+
+error:
+    key_len = NULL;
+    key_data = NULL;
+    DEBUG("[persistent storage] error while decoding key: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+
+}
+
+#if PSA_PROTECTED_KEY_COUNT
+/**
+ * @brief   Decode CBOR encoded protected key
+ *
+ * @param   key     Nanocbor value object containing CBOR encoded key data
+ * @param   slot    Key slot to write decoded key to
+ * @return  psa_status_t
+ */
+static psa_status_t psa_decode_protected_key_slot(nanocbor_value_t *key, psa_key_slot_t *slot)
+{
+    int res = 0;
+
+    psa_key_slot_number_t *slot_no = psa_key_slot_get_slot_number(slot);
+    res = nanocbor_get_uint64(key, slot_no);
```

Comment:
why getting the slot_number in the first place if it will be overridden in the next line?

Reply:
Also if I understand it correctly, this only works for protected key slots, right? In that case I would definitely rename it and maybe even change the type to `psa_prot_key_slot_t`.

Reply:
Hmmm the key types `psa_prot_key_slot_t` and `psa_keypair_slot_t` are subtypes of `psa_key_slot_t`, which should only be used by the key slot management.
PSA modules other than the key slot management should not know about those types and I'd rather not start blurring those boundaries.

Reply:
I'd rather not change this in this PR, because it will affect much more code than this. I'll do this separately.